cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/WrappedCollection.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.spark.batch;
+
+import io.cdap.cdap.etl.api.batch.SparkCompute;
+import io.cdap.cdap.etl.api.batch.SparkSink;
+import io.cdap.cdap.etl.api.streaming.Windower;
+import io.cdap.cdap.etl.common.PhaseSpec;
+import io.cdap.cdap.etl.common.RecordInfo;
+import io.cdap.cdap.etl.common.StageStatisticsCollector;
+import io.cdap.cdap.etl.proto.v2.spec.StageSpec;
+import io.cdap.cdap.etl.spark.SparkCollection;
+import io.cdap.cdap.etl.spark.SparkPairCollection;
+import io.cdap.cdap.etl.spark.join.JoinExpressionRequest;
+import io.cdap.cdap.etl.spark.join.JoinRequest;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.api.java.function.PairFlatMapFunction;
+
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Abstract implementation of a {@link SparkCollection} which wraps another instance in order to delay the execution
+ * of a mapping function.
+ *
+ * This is currently used to prevent delay operations until absolutely needed.
+ * @param <T> Type of the wrapped collection records.
+ * @param <U> Type of the output collection records.
+ */
+public abstract class WrappedCollection<T, U> implements SparkCollection<U> {
```

Comment:
Why do we need this refactoring?

Reply:
@sahusanket This class can be combined with `WrappedSQLEngineCollection` as there are no other classes that extend from WrappedCollection.

Reply:
@tivv This class defines based methods for a Spark collection which wraps another collection.

The implementing class `WrappedSQLEngineCollection` contains additional logic to delegate the execution of certain record mapping operations and "re-wrap" these operations. This allows us to delay certain metadata operations until the records are actually needed in Spark for further processing.

There is also the `WrappableCollection` interface which defines classes that can be safely wrapped into a WrappedCollection. For example, the `SQLEngineCollection` class, and again the "wrapped" version `WrappedSQLEngineCollection`.

If you feel strongly, `WrappedCollection` and `WrappedSQLEngineCollection` can be merged, but I would argue that it makes sense to separate them as `WrappedSQLEngineCollection` contains some custom behavior to handle store and multi store tasks.

I think it makes sense to keep these separated to make the code easier to read.

Reply:
If it's a refactoring, can we do it as a separate PR? 
On the other note, all this wrapping and rewrapping gets really complicated and I think that delegate approach with wrapping is not suitable here. I'd consider moving to another model with clear separation of concerns

Reply:
Wrapping effectively acts as a delegate here. We use wrapping to avoid manipulation of Spark records until the processing if this records in Spark is actually needed.

The other approach I see is to have a chain of delegate methods that we will have to invoke on the SparkCollection once the collection is available in Spark, but this seems like a larger change that we should tackle in a later PR.

My goal is to unblock Sanket's work on the SQL Source Pushdown.

Reply:
I mean, we have this strange wrapping and rewrapping for purposes of tracking in CDAP and then on top of it we have rewrapping in SQL Engine. It get very complex for no real reason. I believe before this PR `WrappedCollection` and `WrappedSQLEngineCollection` was same thing, so it's better not to touch it in this PR unless there is a functional reason to do it. This PR is quite big, so if we feel strongly about refactoring, we can extract it into separate PR and talk about it over there

Reply:
Makes sense. @sahusanket Please condense WrappedCollection and WrappedSQLEngineCollection into the same class for now. 

We can tackle a code cleanup/improvements on this area in a later PR